### ISSUE TITLE
Switch Lighthouse and all deps off `ethereum-types`

### ISSUE THREAD

**Issue by michaelsproul** [sigp/lighthouse#6022]
## Description

Presently Lighthouse is using `ethereum-types` for many of our core type definitions like `Hash256`. This is an old Parity crate that is not actively maintained any more, and there is some desire to move Lighthouse to Alloy so that it synergises better with the _new_ Rust Ethereum ecosystem around Reth, forge, etc.

This task would involve:

1. Updating all our external crates like `ethereum_hashing`, `tree_hash`, `ethereum_ssz` so that they do not depend on `ethereum-types`.
2. Updating Lighthouse to depend on the updated versions of the external crates prior to releasing the crates on `crates.io`. We don't want to commit to merging alloy into our external crates until we're sure it's going to work & not be too much hassle. We should use `patch` to point at the feature branches for each crate so that the crate overrides apply globally (e.g. so that we are also patching transitive dependencies on `ethereum_hashing` via `ethereum_ssz` to use the `git` dep).
3. Update API usage within Lighthouse
4. Merge the changes to the external crates and release new minor versions signalling the breaking change.
5. Update the Lighthouse PR to use the released versions of the crates, and then merge that to unstable.

Part 1 is already underway thanks to @KolbyML:

- https://github.com/sigp/tree_hash/pull/14
- https://github.com/sigp/ethereum_ssz/pull/15
- https://github.com/sigp/ssz_types/pull/22
- https://github.com/sigp/ethereum_serde_utils/pull/7

I have merged these branches down into an `alloy` branch on their respective repositories and created new PRs:

- https://github.com/sigp/tree_hash/pull/16
- https://github.com/sigp/ethereum_ssz/pull/21
- https://github.com/sigp/ssz_types/pull/25
- https://github.com/sigp/ethereum_serde_utils/pull/8

## Ideas to minimise the size of the diff

- We should alias `B256` to `Hash256` to avoid a huge amount of unnecessary rename churn. Even if we want to use the "proper" type name eventually, IMO it is better to separate the semantic change of using alloy, from a bunch of hard-to-review rename churn.

**Comment by jimmygchen** [comment]
I don't see direct reference of `ethereum-types` anymore in Lighthouse, has this been completed? @michaelsproul

**Comment by michaelsproul** [comment]
@jimmygchen No, unfortunately we still depend on two versions of `ethereum-types` transitively:

https://github.com/sigp/lighthouse/blob/e5b4983d6baf85770fe4539a565d8a2dd462bc53/Cargo.lock#L3041-L3069

We need to port the `execution_engine_integration` tests from `ethers` to `alloy`. I've tried to do this a few times and gotten bogged down. I tried to get Claude to do it and it also got bogged down. It's definitely do-able, just a bit tedious.

There's also the `deposit_contract` crate that uses `ethers`. We can probably just completely delete it now that we have Electra, but I haven't looked into the full implications of this.

This issue might be a good one for the Bosun team to tackle. CC @macladson @hopinheimer

**Comment by macladson** [comment]
I have written a series of PRs to remove the remaining deps. The diffs are pretty gross individually (since it's just adding a bunch of alloy deps) but you can see the final result here which shows the overall improvement -> #8150 

If we can remove `deposit_contract` entirely that would improve the diff even more since it needs both `alloy-dyn-abi` and `alloy-json-abi`.
